### PR TITLE
[rustc_parse] Forbid `let`s in certain places

### PR DESCRIPTION
**Pull request by c410-f3r**
Currently only forbids in locals to resolve https://github.com/rust-lang/rust/pull/94927#issuecomment-1099605024 but feel free to point any other places.

### CODE REVIEW

**Comment by compiler-errors** [comment]
Sorry for taking so long to review.

I don't think this should be implemented like this. I think we should be explicitly adding code to _allow_ let-chains in `if` and `while` (i.e. make them deny by default), and not exhaustively forbid them in certain places like the RHS of a `let` statement. 

Even if this is the right approach, I think this current implementation is missing a lot of positions where we'd accidentally stabilize parsing `let` as an expression as discussed in https://github.com/rust-lang/rust/pull/94927#issuecomment-1099605024, for example: 

```rust
#[cfg(ignore)]
fn foo() {
    let _ = f(let y = 1);
    |x| let 1 = 2;
    array[let x = y];
}
```

**Review by compiler-errors — CHANGES_REQUESTED** [review]
(whoops, should've left comment here)

**Comment by c410-f3r** [comment]
> Sorry for taking so long to review.
> 
> I don't think this should be implemented like this. I think we should be explicitly adding code to _allow_ let-chains in `if` and `while` (i.e. make them deny by default), and not exhaustively forbid them in certain places like the RHS of a `let` statement.
> 
> Even if this is the right approach, I think this current implementation is missing a lot of positions where we'd accidentally stabilize parsing `let` as an expression as discussed in [#94927 (comment)](https://github.com/rust-lang/rust/pull/94927#issuecomment-1099605024), for example:
> 
> ```rust
> #[cfg(ignore)]
> fn foo() {
>     let _ = f(let y = 1);
>     |x| let 1 = 2;
>     array[let x = y];
> }
> ```

Done. Thanks for the review

File: compiler/rustc_parse/src/parser/expr.rs [hunk 1]
```diff
@@ -2252,6 +2254,13 @@ impl<'a> Parser<'a> {
     /// Parses a `let $pat = $expr` pseudo-expression.
     /// The `let` token has already been eaten.
     fn parse_let_expr(&mut self, attrs: AttrVec) -> PResult<'a, P<Expr>> {
+        if !self.is_in_a_valid_let_ctx {
+            self.struct_span_err(
+                self.prev_token.span,
+                "expected expression, found `let` statement",
+            )
+            .emit();
+        }
```
Comment (c410-f3r) [inline compiler/rustc_parse/src/parser/expr.rs:2263]: Don't know why but returning `Err` here invalidates any following invalid `let`;

```rust
    &let 0 = 0; // Error!

    !let 0 = 0; // Ignored!
```

File: compiler/rustc_parse/src/parser/mod.rs [hunk 2]
```diff
@@ -147,12 +147,15 @@ pub struct Parser<'a> {
...
+    /// Used to track where `let`s are allowed. For example, `if true && let 1 = 1` is valid
+    /// but `[1, 2, 3][let _ = ()]` is not.
+    is_in_a_valid_let_ctx: bool,
 }
 
 // This type is used a lot, e.g. it's cloned when matching many declarative macro rules. Make sure
 // it doesn't unintentionally get bigger.
 #[cfg(all(target_arch = "x86_64", target_pointer_width = "64"))]
-rustc_data_structures::static_assert_size!(Parser<'_>, 328);
+rustc_data_structures::static_assert_size!(Parser<'_>, 336);
```
Comment (c410-f3r) [inline compiler/rustc_parse/src/parser/mod.rs:158]: Let's just say that this is unavoidable

File: src/test/ui/rfc-2497-if-let-chains/feature-gate.rs [hunk 3]
```diff
@@ -48,9 +49,9 @@ fn _macros() {
     }
     #[cfg(FALSE)] (let 0 = 1);
     //~^ ERROR `let` expressions in this position are unstable [E0658]
+    //~| ERROR expected expression, found `let` statement
     use_expr!(let 0 = 1);
     //~^ ERROR no rules expected the token `let`
-    // ^--- FIXME(53667): Consider whether `Let` can be added to `ident_can_begin_expr`.
```
Comment (c410-f3r) [inline src/test/ui/rfc-2497-if-let-chains/feature-gate.rs:53]: No, `let` is not an expression

File: src/test/ui/rfc-2497-if-let-chains/invalid-let-in-a-valid-let-context.rs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+    #[cfg(FALSE)]
+    {
+        if let Some(elem) = _opt && {
+            [1, 2, 3][let _ = ()];
+            true
+        } {
+        }
+    }
+}
```
Comment (c410-f3r) [inline src/test/ui/rfc-2497-if-let-chains/invalid-let-in-a-valid-let-context.rs:17]: I will fix this issue in a following PR

File: compiler/rustc_parse/src/parser/expr.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_parse/src/parser/expr.rs:2263]: It's because of how you're enabling and disabling `is_in_a_valid_let_ctx`. I'll leave a comment in-line.

File: compiler/rustc_parse/src/parser/expr.rs [hunk 5]
```diff
@@ -2239,7 +2239,9 @@ impl<'a> Parser<'a> {
 
     /// Parses the condition of a `if` or `while` expression.
     fn parse_cond_expr(&mut self) -> PResult<'a, P<Expr>> {
+        self.is_in_a_valid_let_ctx = true;
         let cond = self.parse_expr_res(Restrictions::NO_STRUCT_LITERAL, None)?;
+        self.is_in_a_valid_let_ctx = false;
```
Comment (compiler-errors) [inline compiler/rustc_parse/src/parser/expr.rs:2244]: Rename `is_in_a_valid_let_ctx` to `let_expr_allowed`. Then can you pull this into a function, like:

``` rust
fn with_let_expr_allowed(&mut self, allowed: bool, f: impl FnOnce(&mut Self) -> T) -> T {
    let last_let_expr_allowed = std::mem::replace(&mut self.let_expr_allowed, allowed);
    f(self);
    self.let_expr_allowed = last_let_expr_allowed;
}
```

Then do this:
```suggestion
        let cond = self.with_let_expr_allowed(allowed, |self_|  self_.parse_expr_res(Restrictions::NO_STRUCT_LITERAL, None))?;
```

File: compiler/rustc_parse/src/parser/expr.rs [hunk 6]
```diff
@@ -2595,8 +2604,10 @@ impl<'a> Parser<'a> {
             )?;
             let guard = if this.eat_keyword(kw::If) {
                 let if_span = this.prev_token.span;
+                this.is_in_a_valid_let_ctx = true;
```
Comment (compiler-errors) [inline compiler/rustc_parse/src/parser/expr.rs:2607]: Use `with_let_expr_allowed` like I suggested above.

File: src/test/ui/rfc-2294-if-let-guard/feature-gate.rs [hunk 7]
```diff
@@ -58,8 +58,10 @@ fn _macros() {
     }
     use_expr!((let 0 = 1 && 0 == 0));
     //~^ ERROR `let` expressions in this position are unstable
```
Comment (compiler-errors) [inline src/test/ui/rfc-2294-if-let-guard/feature-gate.rs:60]: I'm not a fan of the duplicated errors -- is it possible to suppress the later errors somehow?

File: compiler/rustc_parse/src/parser/expr.rs [hunk 8]
```diff
@@ -2573,11 +2582,11 @@ impl<'a> Parser<'a> {
     }
 
     pub(super) fn parse_arm(&mut self) -> PResult<'a, Arm> {
-        fn check_let_expr(expr: &Expr) -> (bool, bool) {
+        fn check_let(expr: &Expr) -> (bool, bool) {
```
Comment (compiler-errors) [inline compiler/rustc_parse/src/parser/expr.rs:2585]: nit: why rename this?

File: compiler/rustc_parse/src/parser/expr.rs [hunk 9]
```diff
@@ -2573,11 +2582,11 @@ impl<'a> Parser<'a> {
...
             match expr.kind {
                 ExprKind::Binary(_, ref lhs, ref rhs) => {
```
Comment (compiler-errors) [inline compiler/rustc_parse/src/parser/expr.rs:2587]: nit: should this be considering if the binary op is  `&&`? Not exactly sure how this code works and what it's protecting against -- could you leave a comment?

File: compiler/rustc_parse/src/parser/expr.rs (hunk 1 above)
Comment (c410-f3r) [reply compiler/rustc_parse/src/parser/expr.rs:2263]: Already tried
```rust
self.is_in_a_valid_let_ctx = true;
let cond = match self.parse_expr_res(Restrictions::NO_STRUCT_LITERAL, None) {
  Err(err) => {
    self.is_in_a_valid_let_ctx = false;
    return Err(err);
  }
  Ok(elem) => {
    self.is_in_a_valid_let_ctx = false;
    elem
  }
}
```

But the end result was the same 🤷

File: compiler/rustc_parse/src/parser/expr.rs (hunk 6 above)
Comment (compiler-errors) [reply compiler/rustc_parse/src/parser/expr.rs:2607]: Actually, you can probably use `with_let_expr_allowed` by moving it into `parse_expr` to fix the existing case of this code parsing (incorrectly):

```rust
if let a = 1 && { let x = let y = 1; } {}
```

**Review by compiler-errors — CHANGES_REQUESTED** [review]

File: src/test/ui/rfc-2294-if-let-guard/feature-gate.rs (hunk 7 above)
Comment (compiler-errors) [reply src/test/ui/rfc-2294-if-let-guard/feature-gate.rs:60]: Maybe a follow-up.

File: compiler/rustc_parse/src/parser/expr.rs (hunk 8 above)
Comment (c410-f3r) [reply compiler/rustc_parse/src/parser/expr.rs:2585]: I created this function but `let` is not an expression, thus the renaming.

It is better to revert because the change is not related to the PR.

File: src/test/ui/rfc-2294-if-let-guard/feature-gate.rs (hunk 7 above)
Comment (c410-f3r) [reply src/test/ui/rfc-2294-if-let-guard/feature-gate.rs:60]: FWICT, `parse_let_expr` needs to short-circuit instead of emitting errors

File: compiler/rustc_parse/src/parser/expr.rs (hunk 6 above)
Comment (c410-f3r) [reply compiler/rustc_parse/src/parser/expr.rs:2607]: OK, it is more or less what `ast_validation.rs` already does.

> Actually, you can probably use `with_let_expr_allowed` by moving it into parse_expr

Just a random rant. The whole `expr.rs` file, as well as integrating `with_let_expr_allowed`, glues parsing logic with error management and this feature creates hard to understand and hard and maintain code.

Comment (compiler-errors) [reply compiler/rustc_parse/src/parser/expr.rs:2607]: I certainly agree that `expr.rs` is a bit hard to follow.

**Comment by c410-f3r** [comment]
@rustbot label +S-waiting-on-review -S-waiting-on-author

If possible, I would like to tackle https://github.com/rust-lang/rust/pull/97295/files#diff-52ef6fde627bdaac8e79bcf67beac39ef4cd9a2a0bd0ff7e3ec69dd0ce74af94 and duplicated diagnostics in a following PR.

**Comment by compiler-errors** [comment]
Alright, this is a good first step then.

@bors r+

**Comment by c410-f3r** [comment]
Done
